### PR TITLE
Make Book Hotel Project

### PR DESCRIPTION
**Pull request by zunairkhan811**
The Project has been completed according to the Requirements:
The user logs in to the website, only by typing the username (a proper authenticated login is a requirement if your group is made of 5 people).
In the navigation panel, the user can see links to:
Motorcycles/doctors/classes/items that you selected as a theme.
"Reserve" form.
"My reservations".
"Add motorcycle/doctor/class/item that you selected as a theme" (the link is visible to everybody).
"Delete motorcycle/doctor/class/item that you selected as a theme" (the link is visible to everybody).
On the main page, the user can see a [list of motorcycles/doctors/classes/items that you selected as a theme](https://www.behance.net/gallery/26425031/Vespa-Responsive-Redesign/modules/173005577).
When the user selects a specific item, they can see the [details page](https://www.behance.net/gallery/26425031/Vespa-Responsive-Redesign/modules/173005579) with its full description (skip the "Rotate image" button).
In the details page, the user can click the "Reserve" button (in the design you can see the "Configure" button - please replace it with the "Reserve" button).
When the user clicks the "Add item" link in the navigation panel they can see a form for adding a new item.

When the user clicks the "My reservations" link in the navigation panel they can see a list of their reservations (with information about item name, date and city).
Add full documentation for your API.
Additional features - Required only for teams with 5 members (optional for others)
Implement proper user authentication from the front-end to the server.
Make sure that the "Add item" and "Delete item" links are accessible only by users who are admins.
Technical set up
Set up the repository/repositories on GitHub and use Gitflow.
You should use Postgres as your database
Use Rails to create backend API.
Use React & Redux to create frontend UI.
You can choose if you want to set up your project as two separate apps or as one.
The [Creative Commons license of the design](https://creativecommons.org/licenses/by-nc/4.0/) requires that you give appropriate credit to the author. Therefore, you must do it in the README of your project.

Additional features - Required for teams with 3-4 & 5 members (optional for others)
When the user clicks the "Delete item" link in the navigation panel they can see a list of all items with title and "Delete" button.
When the user clicks the "Delete" button, the selected item is marked as removed and does not show on the main list anymore.
To reserve an appointment, the user has to select a date and city (username and selected item are autofilled).
Use the design based on the ["Book a vespa test-ride"](https://www.behance.net/gallery/26425031/Vespa-Responsive-Redesign/modules/173005583) and add all necessary inputs.
The user can also access the "Reserve" page from the navigation panel. In that case only username is autofilled.

### CODE REVIEW

**Review by Meltrust — CHANGES_REQUESTED** [review]
Hi,

Good job so far!
There are some issues that you still need to work on to prepare your project for the final evaluation, but you are almost there!


Main takeaway:

- This is OPTIONAL:  You have included the frontend app as a folder in the backend app:

#### In the industry when we do different apps (a server API and a frontend app to consume it), we use different repos for each app, we never include the other app as a subfolder.  This is because an API server can serve multiple frontend apps, and also for maintainability.

#### Another option would have been to use the `react-rails` gem in order to make a single monolithic app, however, this is not the case, which is good, because the industry is quickly abandoning single monolithic apps.

#### Therefore I suggest that in the future, when you have the time, take out the `frontend-react-app` folder from the backend and put it in a new repo.  It will work just fine.



#### You are really close to finishing the Microverse program!! Keep it up! 👍👍👍

You can do it

### After implementing the requested changes, please submit another review request. ♻️
_Check the comments under the review._







Cheers and Happy coding!👏👏👏

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the previous reviews unless it is requested otherwise._**

------

File: README.md [hunk 1]
```diff
@@ -1,207 +1,210 @@
...
+### Usage
+
+- To run the project, execute the following command:
+
+```sh
+    rails s
 ```
 
-### Usage
+```sh
```
[the remark this answers is not in the document]
Comment (zunairkhan811) [reply README.md:124]: @Meltrust, thanks for your suggestions, I have implemented the changes.

File: README.md [hunk 2]
```diff
@@ -1,207 +1,210 @@
...
 
-### Install
+- Update the database.yml with your database credentials under the development section
 
-Install this project with:
 
-```
- npm install     // git bash terminal
- bundle install  // wsl terminal
+### Usage
```
[the remark this answers is not in the document]
Comment (zunairkhan811) [reply README.md:116]: @Meltrust, thanks for your suggestions, I have implemented the changes.

[the remark this answers is not in the document]
Comment (zunairkhan811) [reply README.md:116]: @Meltrust, thanks for your suggestions, I have implemented the changes.

File: app/controllers/api/v1/reservations_controller.rb [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+          render json: { error: @reservation.errors.full_messages }, status: :unprocessable_entity
+        end
+      rescue StandardError => e
+        render json: { error: e.message }, status: :internal_server_error
+      end
+
+      def destroy
+        @reservation.destroy
+        head :no_content
+      end
```
[the remark this answers is not in the document]
Comment (zunairkhan811) [reply app/controllers/api/v1/reservations_controller.rb:30]: @Meltrust, thanks for your suggestions, I have implemented the changes.

File: frontend-react-app/src/components/AdditemForm.jsx [hunk 4]
```diff
@@ -0,0 +1,213 @@
+import React, { useState } from 'react';
+import axios from 'axios';
+import { useNavigate } from 'react-router-dom';
+import Navbar from './Navbar';
+
+const AddItemForm = () => {
```
[the remark this answers is not in the document]
Comment (zunairkhan811) [reply frontend-react-app/src/components/AdditemForm.jsx]: @Meltrust, thanks for your suggestions, I have implemented the changes.

File: frontend-react-app/src/components/Carousel.jsx [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+      console.error('Error deleting hotel item:', error);
+    }
+  };
+
+  return (
+    <>
+      <div className="carousel-container">
+        <BackButton onClick={prevSlide} />
+        <div className="carousel-cards-container">
+          {items.map((item) => (
```
[the remark this answers is not in the document]
Comment (zunairkhan811) [reply frontend-react-app/src/components/Carousel.jsx:38]: @Meltrust, thanks for your suggestions, I have implemented the changes.

File: frontend-react-app/src/App.js [hunk 6]
```diff
@@ -1,36 +1,59 @@
...
+import AddItemForm from './components/AddItemForm';
+import DeleteHotel from './components/DeleteHotel';
+
 import Splash from './components/splash';
 
-function App() {
+const App = () => {
+  const isAuthenticated = useSelector((state) => state.auth.isAuthenticated);
+  const dispatch = useDispatch();
+  console.log('appauthentication', isAuthenticated);
```
[the remark this answers is not in the document]
Comment (zunairkhan811) [reply frontend-react-app/src/App.js:21]: @Meltrust, thanks for your suggestions, I have implemented the changes.

**Review by Meltrust — CHANGES_REQUESTED** [review]
Hi,

Good job so far!
There are some issues that you still need to work on to prepare your project for the final evaluation, but you are almost there!






### To highlight:

-  Almost all tests are now passing✔️
-  Nice code organization ✔️ 
-  The app is working well✔️ 
-  Readme has been vastly improved✔️ 
-  Superfluous test folder has been removed✔️
-  Webpack compilation errors have been fixed ✔️ 
-  The app now works even if there are no items on the database ✔️ 
-  Error handling has been added to the destroy method in the controller ✔️ 



#### You are really close to finishing the Microverse program!! Keep it up! 👍👍👍

You can do it

### After implementing the requested changes, please submit another review request. ♻️
_Check the comments under the review._







Cheers and Happy coding!👏👏👏

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the previous reviews unless it is requested otherwise._**

------

File: README.md [hunk 7]
```diff
@@ -1,207 +1,235 @@
...
 
+**This project is deployed by the author, no permission for deployment by any other client.**
 
-<p align="right">(<a href="#readme-top">back to top</a>)</p>
+### Tests
+
+- To run the **tests** of project, execute the following command in root folder:
+```sh
+    --open your gitbash terminal--
+    rspec ./spec/integration
```
[the remark this answers is not in the document]
Comment (zunairkhan811) [reply README.md:153]: Hi @Meltrust, thanks, Your suggestion has been implemented 😃

**Review by Meltrust — APPROVED** [review]
# **Wow, you did it 🎉**
Brilliant
## Thank you for the changes implemented 💪 🥇 ㊗️ 
#### _Unless you want to add more features, go ahead to your final presentation ⏩ ⏩ ⏩_ 
## You are about to finish the Microverse program.  You have come a long way!!!




### Good luck in the software industry!!  I'll see you there. ✨
### **Congratulations!!!!!!**  🎉
applause

### To highlight

- All tests are now passing ✔️ 
- Superb job ✔️  



Cheers and Happy coding!👏👏👏


------

**Pull request merged**